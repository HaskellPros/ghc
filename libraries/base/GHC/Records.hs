{-# LANGUAGE AllowAmbiguousTypes
           , DataKinds
           , ExplicitForAll
           , FunctionalDependencies
           , KindSignatures
           , MultiParamTypeClasses
           , NoImplicitPrelude
           , PolyKinds
           , ScopedTypeVariables
           , TypeApplications
  #-}

-----------------------------------------------------------------------------
-- |
-- Module      :  GHC.Records
-- Copyright   :  (c) Adam Gundry 2015
-- License     :  see libraries/base/LICENSE
--
-- Maintainer  :  cvs-ghc@haskell.org
-- Stability   :  internal
-- Portability :  non-portable (GHC extensions)
--
--
-----------------------------------------------------------------------------

module GHC.Records
       ( HasField(..)
       ) where

import GHC.Base ( Symbol )

class HasField (x :: k) r a | x r -> a where
  fromLabel :: r -> a
