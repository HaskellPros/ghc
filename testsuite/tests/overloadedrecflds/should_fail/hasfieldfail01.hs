{-# LANGUAGE DataKinds, MagicHash, TypeFamilies, TypeApplications #-}

import HasFieldFail01_A (T(MkT))

import GHC.Records (HasField(..))

-- This should fail to solve the HasField constraint, because foo is
-- not in scope.
main = print (fromLabel @"foo" (MkT 42) :: Int)
