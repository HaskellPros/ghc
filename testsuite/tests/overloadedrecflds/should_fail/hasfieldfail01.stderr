[1 of 2] Compiling HasFieldFail01_A ( HasFieldFail01_A.hs, HasFieldFail01_A.o )
[2 of 2] Compiling Main             ( hasfieldfail01.hs, hasfieldfail01.o )

hasfieldfail01.hs:9:15: error:
    • No instance for (HasField "foo" T Int)
        arising from a use of ‘fromLabel’
    • In the first argument of ‘print’, namely
        ‘(fromLabel @"foo" (MkT 42) :: Int)’
      In the expression: print (fromLabel @"foo" (MkT 42) :: Int)
      In an equation for ‘main’:
          main = print (fromLabel @"foo" (MkT 42) :: Int)
